vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java

```diff
@@ -97,7 +100,7 @@
     public final static List<IAlgebraicRewriteRule> buildPathStepNormalizationRuleCollection() {
         List<IAlgebraicRewriteRule> normalization = new LinkedList<IAlgebraicRewriteRule>();
         normalization.add(new SetVariableIdContextRule());
-
+        normalization.add(new InferTypesRule());
```

Comment:
InferTypesRule: I put it in the beginning of the rules set because in order for the group by feature to work correctly, the type of the expression should be set in the beginning. 

All the other rules are about group by and an operator that is above or inside the group by operator. I put them in the corresponding positions because these were the points that the suitable operators were at the position that could activate the rule. 

I have put comments in the new pull request inside the new rules.